lib/xml_builder.ex

```diff
@@ -122,85 +88,10 @@ defmodule XmlBuilder do
-    do: format({nil, nil, string}, level)
-
-  defp format(list, level) when is_list(list),
-    do: list |> Enum.map(&format(&1, level)) |> Enum.intersperse("\n")
-
-  defp format({nil, nil, name}, level) when is_bitstring(name),
-    do: [indent(level), to_string(name)]
-
-  defp format({name, attrs, content}, level) when is_blank_attrs(attrs) and is_blank_list(content),
-    do: [indent(level), '<', to_string(name), '/>']
-
-  defp format({name, attrs, content}, level) when is_blank_list(content),
-    do: [indent(level), '<', to_string(name), ' ', format_attributes(attrs), '/>']
-
-  defp format({name, attrs, content}, level) when is_blank_attrs(attrs) and not is_list(content),
-    do: [indent(level), '<', to_string(name), '>', format_content(content, level+1), '</', to_string(name), '>']
-
-  defp format({name, attrs, content}, level) when is_blank_attrs(attrs) and is_list(content),
-    do: [indent(level), '<', to_string(name), '>', format_content(content, level+1), '\n', indent(level), '</', to_string(name), '>']
-
-  defp format({name, attrs, content}, level) when not is_blank_attrs(attrs) and not is_list(content),
-    do: [indent(level), '<', to_string(name), ' ', format_attributes(attrs), '>', format_content(content, level+1), '</', to_string(name), '>']
-
-  defp format({name, attrs, content}, level) when not is_blank_attrs(attrs) and is_list(content),
-    do: [indent(level), '<', to_string(name), ' ', format_attributes(attrs), '>', format_content(content, level+1), '\n', indent(level), '</', to_string(name), '>']
-
-  defp tree_node(element_spec),
-    do: element(element_spec)
-
-  defp format_content(children, level) when is_list(children),
-    do: ['\n', Enum.map_join(children, "\n", &format(&1, level))]
-
-  defp format_content(content, _level),
-    do: escape(content)
-
-  defp format_attributes(attrs),
-    do: Enum.map_join(attrs, " ", fn {name,value} -> [to_string(name), '=', quote_attribute_value(value)] end)
-
-  defp indent(level),
-    do: String.duplicate("\t", level)
-
-  defp quote_attribute_value(val) when not is_bitstring(val),
-    do: quote_attribute_value(to_string(val))
-
-  defp quote_attribute_value(val) do
-    double = String.contains?(val, ~s|"|)
-    single = String.contains?(val, "'")
-    escaped = escape(val)
-
-    cond do
-      double && single ->
-        escaped |> String.replace("\"", "&quot;") |> quote_attribute_value
-      double -> "'#{escaped}'"
-      true -> ~s|"#{escaped}"|
-    end
-  end
-
-  defp escape({:cdata, data}) do
-    ["<![CDATA[", data, "]]>"]
-  end
-
-  defp escape(data) when not is_bitstring(data),
-    do: escape(to_string(data))
+  def generate(any, opts \\ [formatter: XmlBuilder.Formatters.Default]),
+    do: apply(opts[:formatter] || default(), :format, [any, 0]) |> IO.chardata_to_string
 
-  defp escape(string) do
-    string
-    |> String.replace(">", "&gt;")
-    |> String.replace("<", "&lt;")
-    |> replace_ampersand
-  end
+  defp default,
```

Comment:
Since this commit supports `:indented`-like short names for formatters, nested in `XmlBuilder.Formatters` module, it’s now `safe_formatter`.

The formatter’s module name is already accepted by `XmlBuilder.generate/2` and while passed it takes precedence over application config. Please let us keep an ability to set the default parameter one for all in the config: it’s the lowest precedence source for the value, but many usecases just do not need the formatter to be re-set on the fly and the ability to specify it in `config` saves a ton of typing.

`XmlBuilder.doc/1` was not granted with the ability to accept `formatter` parameter since it would introduce breaking changes and I am not brave enough to introduce breaking changes in PR.

Please do it youself if you consider it’s necessary.